<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید. 
</font>
</p>

In [1]:
import pandas as pd
data= pd.read_csv('NBA2000-2009.csv')
data.tail()

PLAYER TEAM  YEAR  GP   MIN  PTS  FGM  FGA  3PM  3PA  FTM  FTA  \
1825    Marcin Gortat  ORL  2009  81  13.4  3.6  1.5  2.8  0.0  0.0  0.6  0.9   
1826      Jeff Teague  ATL  2009  71  10.1  3.2  1.3  3.2  0.1  0.5  0.6  0.7   
1827      Kurt Thomas  MIL  2009  70  15.0  3.0  1.4  3.0  0.0  0.0  0.2  0.2   
1828  DeMarre Carroll  MEM  2009  71  11.2  2.9  1.2  3.1  0.0  0.1  0.5  0.7   
1829     Joel Anthony  MIA  2009  80  16.5  2.7  1.0  2.0  0.0  0.0  0.8  1.2   

      OREB  DREB  AST  STL  BLK  TOV  
1825   1.3   2.9  0.2  0.2  0.9  0.6  
1826   0.1   0.9  1.7  0.5  0.2  0.7  
1827   1.2   3.0  0.7  0.4  0.7  0.7  
1828   0.7   1.4  0.5  0.4  0.1  0.3  
1829   1.5   1.6  0.2  0.3  1.4  0.6

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد. 
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

Nbaex["YEAR"].value_counts()
Nbaex[Nbaex['PLAYER']=='Antonio MeDyess']
print(data.columns.tolist())



In [8]:
import pandas as pd
data= pd.read_csv('NBA2000-2009.csv')
data['REB']=data['DREB'] + data['OREB']
double_double_list =[]
for index ,row in data.iterrows():
    count=0
    if row['PTS'] >= 10 :
        count += 1
    if row['AST'] >= 10 :
        count += 1 
    if row['REB'] >= 10 :
        count += 1  
    if row['STL'] >= 10 :
        count += 1
    if row['BLK'] >= 10 :
        count += 1
    if count >= 2:
        double_double_list.append({
            'PLAYER':row ['PLAYER'],
            'YEAR':row ['YEAR'],
            'PTS':row ['PTS'],
            'AST':row ['AST'],
            'REB':row ['REB'],
            'STL':row ['STL'],
            'BLK':row ['BLK'],
        })

double_double =pd.DataFrame(double_double_list)
double_double =double_double.sort_values(['PLAYER','YEAR'],ascending = [ True ,True ])
print(double_double)
       



             PLAYER  YEAR   PTS   AST   REB  STL  BLK
61     Al Jefferson  2007  21.0   1.4  11.2  0.9  1.5
15     Andre Miller  2001  16.5  10.9   4.6  1.6  0.4
58   Antawn Jamison  2007  21.4   1.5  10.2  1.3  0.4
7     Antonio Davis  2000  13.7   1.4  10.1  0.3  1.9
4   Antonio McDyess  2000  20.8   2.1  12.0  0.6  1.5
..              ...   ...   ...   ...   ...  ...  ...
87      Troy Murphy  2009  14.6   2.1  10.2  1.0  0.5
68   Tyson Chandler  2007  11.8   1.0  11.8  0.6  1.1
29    Zach Randolph  2003  20.1   2.0  10.5  0.8  0.5
50    Zach Randolph  2006  23.6   2.2  10.1  0.8  0.2
79    Zach Randolph  2009  20.8   1.8  11.8  1.0  0.4

[88 rows x 7 columns]


<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

In [11]:
import pandas as pd
data= pd.read_csv('NBA2000-2009.csv')
data['Missed FT']= data['FTA'] - data['FTM']
data['Missed FG']= data['FGA'] - data['FGM']
#(PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT)
data['VALUE'] = (data['PTS'] + data['OREB'] + data['DREB'] + 
                 data['AST'] + data['STL'] + data['BLK']) - \
                (data['TOV'] + data['Missed FG'] + data['Missed FT'])

player_avg_value = data.groupby('PLAYER')['VALUE'].mean().reset_index()
player_avg_value ['VALUE']=player_avg_value['VALUE'].round(2)
best_player =player_avg_value[['PLAYER','VALUE']]

\
best_player = best_player.sort_values(['VALUE','PLAYER'],ascending =[True,False])


print(best_player)



                   PLAYER  VALUE
478          Sam Mitchell   3.15
415  Nikoloz Tskitishvili   3.30
65       Brian Scalabrine   3.30
512            Steve Kerr   3.60
432          Quincy Douby   3.60
..                    ...    ...
169           Dwyane Wade  26.12
495      Shaquille O'Neal  26.55
98             Chris Paul  26.57
335          LeBron James  27.90
316         Kevin Garnett  29.74

[571 rows x 2 columns]


<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [17]:
team_pts_year = data.groupby(['YEAR', 'TEAM'])['PTS'].sum().reset_index()

max_pts_teams = team_pts_year.loc[team_pts_year.groupby('YEAR')['PTS'].idxmax()]

max_pts_teams['PTS'] = max_pts_teams['PTS'].round(2)

max_pts_teams = max_pts_teams.sort_values('YEAR')

max_PTS_of_year = max_pts_teams[['YEAR', 'TEAM', 'PTS']].reset_index(drop=True)

print(max_PTS_of_year)


   YEAR TEAM    PTS
0  2000  SAC  100.2
1  2001  LAL   95.9
2  2002  ORL   97.0
3  2003  DEN   96.0
4  2004  BOS  106.2
5  2005  WAS   96.8
6  2006  PHX  105.4
7  2007  DEN  110.7
8  2008  DEN   97.3
9  2009  OKC  102.5


<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [18]:
import zipfile
import os

if not os.path.exists(os.path.join(os.getcwd(), 'NBA.ipynb')):
    %notebook -e NBA.ipynb

double_double.to_csv('double_double.csv', index=False)
best_player.to_csv('best_player.csv', index=False)
max_PTS_of_year.to_csv('max_PTS_of_year.csv', index=False)

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

file_names = ['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
compress(file_names)

File Paths:
['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
